codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -147,32 +160,210 @@ public void onHttpServerUpgrade(Http2Settings settings) throws Http2Exception {
         connection().remote().createStream(HTTP_UPGRADE_STREAM_ID).open(true);
     }
 
-    @Override
-    public void channelActive(ChannelHandlerContext ctx) throws Exception {
-        // The channel just became active - send the connection preface to the remote endpoint.
-        sendPreface(ctx);
-        super.channelActive(ctx);
+    private abstract class Http2ByteToMessageDecoder extends ByteToMessageDecoder {
+        @Override
+        public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+            try {
+                ChannelFuture future = ctx.newSucceededFuture();
+                final Collection<Http2Stream> streams = connection().activeStreams();
+                for (Http2Stream s : streams.toArray(new Http2Stream[streams.size()])) {
+                    closeStream(s, future);
+                }
+            } finally {
+                try {
+                    encoder().close();
+                } finally {
+                    decoder().close();
+                }
+            }
+            super.channelInactive(ctx);
+        }
+
+        /**
+         * Handles {@link Http2Exception} objects that were thrown from other handlers. Ignores all other exceptions.
+         */
+        @Override
+        public final void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+            if (getEmbeddedHttp2Exception(cause) != null) {
+                // Some exception in the causality chain is an Http2Exception - handle it.
+                onException(ctx, cause);
+            } else {
+                super.exceptionCaught(ctx, cause);
+            }
+        }
+
+        @Override
+        public final void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+            Http2ConnectionHandler.this.userEventTriggered(ctx, evt);
+            super.userEventTriggered(ctx, evt);
+        }
+
+        /**
+         * Determine if the HTTP/2 connection preface been sent.
+         */
+        public boolean prefaceSent() {
+            return true;
+        }
+    }
+
+    private final class PreInitializedByteToMessageDecoder extends Http2ByteToMessageDecoder {
+        private ByteBuf clientPrefaceString;
+        private boolean prefaceSent;
+
+        @Override
+        public boolean prefaceSent() {
+            return prefaceSent;
+        }
+
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+            try {
+                if (readClientPrefaceString(in)) {
+                    // After the preface is read, it is time to hand over control to the post initialized decoder.
+                    Http2ConnectionHandler.this.byteDecoder = new PostInitializedByteToMessageDecoder();
+                    ctx.pipeline().replace(this, INTERNAL_HANDLER_NAME, Http2ConnectionHandler.this.byteDecoder);
```

Comment:
That is a great question. In a sense yes it is. [ByteToMessageDecoder.handlerRemoved](https://github.com/netty/netty/blob/master/codec/src/main/java/io/netty/handler/codec/ByteToMessageDecoder.java#L201) does read to another buffer.  I was trying to retain/reuse the existing buffer at one point but realized this was going on under the hood.  I haven't put any effort into mitigating this yet, but it may be worth exploring this now....